packages/examples/src/diffie-hellman/diffie-hellman.ts

```diff
@@ -0,0 +1,158 @@
+import { Choreography, Located } from "@choreography-ts/core";
+import { ExpressBackend } from "@choreography-ts/backend-express";
+
+// Typescript implementation of the Diffie-Hellman key exchange algorithm
+// Ported from the HasChor implementation here: https://github.com/gshen42/HasChor/blob/42ae1ef9a500dadd82f0cfe5dee3c2aa631d8f4d/examples/diffiehellman/Main.hs
+
+//////////////////////
+// Helper functions //
+/////////////////////
+
+// Return all divisors of `x`
+const divisors = (x: number): number[] => {
+  let divs: number[] = [1, x];
+  for (let y = 2; y <= x / 2; y++) {
+    if (x % y == 0) divs.push(y);
+  }
+  return divs;
+};
+
+// Checks if input number is prime
+const isPrime = (x: number): boolean => {
+  let divs: number[] = divisors(x);
+  if (divs.length == 2) return true;
+  return false;
+};
+
+// Prime number generator
+const primeNums = function* () {
+  let y = 1;
+  while (y++ > 0) {
+    if (isPrime(y)) yield y;
+  }
+};
+
+// Return random integer in range [min, max]
+// https://stackoverflow.com/a/7228322
+const randInRange = (min: number, max: number) =>
+  Math.floor(Math.random() * (max - min + 1) + min);
+
+///////////////////////////////////
+// Diffie-Hellman implementation //
+///////////////////////////////////
+const locations = ["alice", "bob"];
+export type Location = (typeof locations)[number];
+
+export const diffieHellman = <A extends Location, B extends Location>(
+  a: A,
+  b: B
+) => {
+  const choreo: Choreography<
+    Location,
+    [Located<boolean, A>], // Specify boolean argument specifically for location "A" ("A|B" instead would mean for both)
+    [Located<number, A>, Located<number, B>]
+  > = async ({ locally, comm }, [arg]) => {
+    // The `arg` specifies whether `a` should wait to initiate key exchange
+
+    // Wait for alice to start key exchange (if desired)
+    await locally(a, (unwrap: any) => {
+      console.log("Press enter to begin key exchange...");
+      // Check for key input on stdin in node: https://stackoverflow.com/a/72906729
+      let status: boolean = !unwrap(arg);
+      if (!status) {
+        // If waiting is desired
```

Comment:
You should not have to busy-wait here. Can't you just return a new promise that resolves when the user press some key?

Reply:
Yeah, I was trying that, but the issue was that the EventListener in Node can't return any value outside the event handler, so the only way I could think of detecting the keypress was to wait and repeatedly check whether it's been pressed using `setInterval`. I'll look into it a bit more though since I too think I shouldn't have to busy-wait

Reply:
Ok yeah you're right, I don't have to busy wait and can just return a Promise. I just fixed it and pushed the changes.